docs/contributing/999-rfcs/2024-06-02-ts-client.md

```diff
@@ -0,0 +1,297 @@
+  // version: "0.0.0",
+  // outdir: join(@dirname, "dist"),
+);
+```
+
+To lift a resource into the client to make it accessible to consumers, add a `@lift` expression:
+
+```wing
+bring cloud;
+
+let bucket = new cloud.Bucket();
+
+let client = new clients.TypeScript(
+  name: "@acme/infra-client",
+  lifts: @lift({ bucket: [put] }),
+);
+```
+
+In order to access the lifted objects at runtime, the TypeScript client needs to know how it should obtain the initialization data for instantiating the inflight clients (e.g. resource ARNs and other preflight state). This information is stored in client.context in preflight.
+When the client is used later (inflight), the environment variable `WING_CONTEXT` is expected to the same value previously stored in client.context.
+
+The context is just a string value (base64 encoded JSON), so you can store it in any way that is convenient for your deployment process.
+
+For example, it could be stored in a bucket:
+
+```wing
+bring cloud;
+bring clients;
+bring fs;
+
+let client = new clients.TypeScript(
+  name: "@acme/infra-client",
+);
+
+// The deployer of the client host will need to retrieve/link this data later
+let clientContext = new cloud.Bucket();
+clientContext.addObject("ctx", client.context)
+```
+
+Regardless of the approach used to resolve the context, using the generated client will work the same way:
+
+```ts
+import { Client } from "@acme/infra-client";
+
+// Will use the environment variable WING_CONTEXT to resolve the context
+const client = new Client();
+
+// Optionally, you can pass the context directly to bypass the environment variable
+// const client = new Client({ context: "..." });
+
+await client.bucket.put("key", "value");
+```
+
+The client code is the same as the generated javascript as if it were used within wing itself!
+
+## Implementation Details
+
+### Client context data
+
+The client context data is a JSON object that contains all the necessary information to connect to the resources.
+It's read through the environment variable `WING_CONTEXT` and is expected to be a base64 encoded.
+
+For now, the only information that needs to be stored are environment variables.
+This corresponds to the only API available on `IInflightHost`, which is `addEnvironment`, which is how these values are added:
+
+```json
+{
+  "version": "1",
+  "env": {
+    "BUCKET_HANDLE_a107cec5": "abcxyz"
+  }
+}
+```
+
+The client itself is effectively a proxy of an unseen/unavailable `IInflightHost`, so any expressivity that interface has must be handled by the client.
+
+### Permissions
+
+The client does not directly/automatically handle permissions needed for the arbitrary host it could run on.
+There is no expectation that the deployer of the host is also using wing so there is no shared system to transfer permission knowledge.
+
+For example, when using
+
+```wing
+let client = new clients.TypeScript(
+  name: "@acme/infra-client",
+  lifts: @lift({ bucket: [put] }),
```

Comment:
1. How does syntax work if I want to lift `myArray[0]` for example?
2. How do I lift multiple objects?

Reply:
Both of these questions/answers are part of the original syntax. We can improve them but I think that belongs in its own discussion. A benefit of reusing `lift` here is that any improvements to it will benefit its original inflight usage and the usage proposed here.

> How does syntax work if I want to lift myArray[0] for example?

It can be used as is:

```wing
@lift({ buckets[0]: [put] })
```

If the resulting name isn't good, `let` can be used

```wing
let bucket = buckets[0];
@lift({ bucket: [put] })
```

> How do I lift multiple objects?

The existing syntax supports this already

```wing
@lift({ bucket1: [put], bucket2: [get] })
```

Reply:
Ok, we need to figure out a better solution. This is still not good.

Reply:
Will that block this RFC?

Reply:
No, it's for the `@lift` RFC :-)